### PR TITLE
Get relative image pathing working for unit test (GLSL, OSL, MDL)

### PR DESCRIPTION
**Pull request by bernardkwok**
Update #840

- Make sure to add in root libraries/ path for material library examples.
- Modify parquet and add brick so that image inputs are exposed. This is required for introspection to find the paths so that they can be converted to absolute paths properly.
(Note the oslc unit test path does the patching but would not handle parquet and brick since there are no exposed uniforms).
- Fix relative paths for MDL when running `mdlc` by adding in mtlx path and root libraries path.

The alternative is still to provide addition image paths but patching provide portability without this.

Unit test results. Note that the nodegraph render also works once exposed at the nodedef level.
image

For example the nodegraph shader will have relative paths:
```code
surface parquet_instance_nodegraph_parquet_out
(
    string parquet_03_base_color_image = "adsk/textures/parquet_03/2k/parquet_03_basecolor_2k_acescg.exr"
    [[
        string widget = "filename"
    ]],
    string parquet_03_normalmap_image = "adsk/textures/parquet_03/2k/parquet_03_normal_2k_raw.exr"
    [[
        string widget = "filename"
    ]],
    string parquet_03_roughness_image = "adsk/textures/parquet_03/2k/parquet_03_roughness_2k_raw.exr"
    [[
        string widget = "filename"
    ]],
    output surfaceshader parquet_out = 0
)
```
with a absolute path overrides applied during usage:
```code
  <ShaderGroup>
      string parquet_03_base_color_image "D:/Work/materialx/MaterialX-adsk-public-master/build_public_rel/source/MaterialXTest/libraries/adsk/textures/parquet_03/2k/parquet_03_basecolor_2k_acescg.exr";
string parquet_03_normalmap_image "D:/Work/materialx/MaterialX-adsk-public-master/build_public_rel/source/MaterialXTest/libraries/adsk/textures/parquet_03/2k/parquet_03_normal_2k_raw.exr";
string parquet_03_roughness_image "D:/Work/materialx/MaterialX-adsk-public-master/build_public_rel/source/MaterialXTest/libraries/adsk/textures/parquet_03/2k/parquet_03_roughness_2k_raw.exr";
;
      shader parquet_instance_nodegraph_parquet_out inputShader;
      shader closure_passthrough outputShader;
      connect inputShader.parquet_out outputShader.Cin;
   </ShaderGroup>
```

### CODE REVIEW

File: libraries/adsk/shaders/brick/ND_brick_02_shader.mtlx [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0"?>
+<materialx version="1.37">
+  <nodedef name="ND_brick_02_surfaceshader" node="brick_02" nodegroup="shader" version="1.0" isdefaultversion="true" namespace="adsk">
+     <parameter name="ao_image" type="filename" value="adsk/textures/bricks_02/2k/bricks_02_ao_2k_raw.exr" />
```
Comment (bernardkwok) [inline libraries/adsk/shaders/brick/ND_brick_02_shader.mtlx:4]: Expose at this level so it shows up as uniforms which can be queried and modified if needed.

File: libraries/adsk/shaders/brick/ND_brick_02_shader.mtlx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     <parameter name="color_image" type="filename" value="adsk/textures/bricks_02/2k/bricks_02_basecolor_2k_acescg.exr" colorspace="acescg" />
+     <parameter name="normalmap_image" type="filename" value="adsk/textures/bricks_02/2k/bricks_02_normal_2k_raw.exr" />
+     <parameter name="roughness_image" type="filename" value="adsk/textures/bricks_02/2k/bricks_02_roughness_2k_raw.exr" />
+     <output name="out" type="surfaceshader" value="" />
+  </nodedef>
+  <nodegraph name="NG_brick_02_surfaceshader" nodedef="adsk:ND_brick_02_surfaceshader" namespace="adsk">
```
Comment (bernardkwok) [inline libraries/adsk/shaders/brick/ND_brick_02_shader.mtlx:10]: For the most part this just becomes a template that could be reused. as there isn't really really anything that makes it a "brick". The thing is I'm not sure how to make this reusable unless it's a definition itself which is the current possible thought.

File: source/MaterialXTest/MaterialXRender/RenderUtil.cpp [hunk 3]
```diff
@@ -296,6 +296,8 @@ bool ShaderRenderTester::validate(const mx::FilePathVec& testRootPaths, const mx
             renderableSearchTimer.endTimer();
 
             mx::FileSearchPath imageSearchPath(dir);
+            imageSearchPath.append(searchPath);
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXRender/RenderUtil.cpp:299]: Missing setting the root of "libraries" as a search path for images.

**Comment by ashwinbhat** [comment]
Just want to clarify, does this mean that in "Value" for filename will point to fully resolved paths? How will fileprefix get handled?

**Comment by bernardkwok** [comment]
> Just want to clarify, does this mean that in "Value" for filename will point to fully resolved paths? How will fileprefix get handled?

This is just one instance of usage. Anyone examining a Shader if they want to access the inputs then as a "recommended" practice the nodedef should have these as exposed interfaces. The actual resolving of the name is up to the integration. In this case it performs relative to absolute path resolves but any other name resolvers can be applied include token handling.

The 3rd alternative is to allow for name resolves during shadergen but then this would tie shader code inputs to the pathing and resolvers at time of shader generation which make it much less flexible.

File: libraries/adsk/adsklib/colorcorrect.mtlx [hunk 4]
```diff
@@ -3,8 +3,8 @@
 
    <!-- Sample color correct interface. Versioned and namespaced -->
    <nodedef name="ND_colorcorrect" node="colorcorrect" nodegroup="adjustment" version="1.0" isdefaultversion="true" namespace="adsk">
-      <parameter name="color" type="color3" value="0.5, 0.5, 0.5" />
-      <parameter name="alpha" type="float" value="1" />
+      <input name="color" type="color3" value="0.5, 0.5, 0.5" />
```
Comment (bernardkwok) [inline libraries/adsk/adsklib/colorcorrect.mtlx:6]: This failed for code gen as we don't allow upstream traversal from a parameter when we create an instance of this, making this node pretty well useless for anything other than modifying the constants in the definition.

File: source/MaterialXTest/MaterialXGenMdl/GenMdl.cpp [hunk 5]
```diff
@@ -140,9 +140,11 @@ void MdlShaderGeneratorTester::compileSource(const std::vector<mx::FilePath>& so
     mdlcCommand += " -p \"" + coreModulePath.asString() + "\"";
     mdlcCommand += " -p \"" + coreModulePath2.asString() + "\"";
     mdlcCommand += " -p \"" + moduleToTestPath.asString() + "\"";
+    mdlcCommand += " -p \"" + moduleToTestPath.getParentPath().asString() + "\"";
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXGenMdl/GenMdl.cpp:143]: Add mtlx file path and root libraries path to fix errors due to missing images.

File: source/MaterialXTest/MaterialXGenShader/GenShaderUtil.cpp [hunk 6]
```diff
@@ -833,22 +838,7 @@ void ShaderGeneratorTester::validate(const mx::GenOptions& generateOptions, cons
...
-                        {
-                            path.createDirectory();
-                        }
-                        path = path / _shaderGenerator->getTarget();
-                        if (!path.exists())
-                        {
-                            path.createDirectory();
-                        }
-                        
+
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXGenShader/GenShaderUtil.cpp:841]: Place mdl shaders in same location as shaders for other languages.

File: source/MaterialXTest/MaterialXGenShader/GenShaderUtil.cpp [hunk 7]
```diff
@@ -788,6 +792,7 @@ void ShaderGeneratorTester::validate(const mx::GenOptions& generateOptions, cons
 
                 mx::string elementName = mx::replaceSubstrings(namePath, pathMap);
                 elementName = mx::createValidName(elementName);
+                elementName = mx::replaceSubstrings(elementName, filenameRemap);
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXGenShader/GenShaderUtil.cpp:795]: Need to do this to remove namespace separators which are valid in an Element name but bad as a file name.

**Pull request merged**